k2/csrc/fsa_algo.cu

```diff
@@ -247,4 +247,11 @@ void ArcSort(Fsa *fsa) {
   SortSublists<Arc, ArcComparer>(fsa);
 }
 
+void ArcSort(Fsa &src, Fsa *dest, Array1<int32_t> *arc_map /*= nullptr*/) {
+  K2_CHECK_GE(src.NumAxes(), 2) << "Input FSA is empty!";
+  *dest = src.Clone();
+  if (arc_map) *arc_map = Array1<int32_t>(src.Context(), src.values.Dim());
```

Comment:
use `if(arc_map != nullptr)` as we have discussed the style in if statement?

Reply:
We shouldn't do duplicate or unnecessary work.  Axes whose row_splits and row_ids are unchanged should be copied, not cloned; and we shouldn't copy arrays twice.  (The function that sorts the arc should accept input-array != output-array).

Reply:
.. actually, in this case the output has the same shape as the input; we don't need to do Clone() at all.
There is an implicit assumption that row-ids and row-splits won't be modified in-place, so we freely share them between objects.

Reply:
> We shouldn't do duplicate or unnecessary work. Axes whose row_splits and row_ids are unchanged should be copied

Sharing the underlying `shape` memory of two FSAs has the risk that changing one FSA will affect another.

---


> The function that sorts the arc should accept input-array != output-array)

Agreed. We should check that `&src  != dest`.

Reply:
What I meant is the function that sorts the arcs should accept both input-array == output-array and input-array != output-array.  This is how sorts generally work.

Reply:
Regarding: `Sharing the underlying shape memory of two FSAs has the risk that changing one FSA will affect another.`... there is intended to be a convention that memory used in RaggedShape is `read-only` (unless you just created it).  I.e. you should generally not modify the memory of a RaggedShape unless you were specifically asked to.

Reply:
Oh I see now what you meant.  
I think it's not necessary to require that &src != dest in ArcSort() and similar functions.
The way we generally do this is do all the work internally without modifying *dest;
we create an Fsa (or its components) internally,
and then right at the end, do something like
` *dest = ans;`
or
`*dest = Ragged<Arc>(shape, arcs);`
(which just copies the metadata).  So actually it won't matter if dest == &src.

Reply:
I see.